README.rst
Python Client for Cloud Optimization API
========================================

|stable| |pypi| |versions|

`Cloud Optimization API`_: is a managed routing service that takes your list of orders, vehicles, constraints, and objectives and returns the most efficient plan for your entire fleet in near real-time.

- `Client Library Documentation`_
- `Product Documentation`_

.. |stable| image:: https://img.shields.io/badge/support-stable-gold.svg
   :target: https://github.com/googleapis/google-cloud-python/blob/main/README.rst#stability-levels
.. |pypi| image:: https://img.shields.io/pypi/v/google-cloud-optimization.svg
   :target: https://pypi.org/project/google-cloud-optimization/
.. |versions| image:: https://img.shields.io/pypi/pyversions/google-cloud-optimization.svg
   :target: https://pypi.org/project/google-cloud-optimization/
.. _Cloud Optimization API: https://cloud.google.com/optimization/docs
.. _Client Library Documentation: https://cloud.google.com/python/docs/reference/cloudoptimization/latest
.. _Product Documentation:  https://cloud.google.com/optimization/docs

Quick Start
-----------

In order to use this library, you first need to go through the following steps:

1. `Select or create a Cloud Platform project.`_
2. `Enable billing for your project.`_
3. `Enable the Cloud Optimization API.`_
4. `Setup Authentication.`_

.. _Select or create a Cloud Platform project.: https://console.cloud.google.com/project
.. _Enable billing for your project.: https://cloud.google.com/billing/docs/how-to/modify-project#enable_billing_for_a_project
.. _Enable the Cloud Optimization API.:  https://cloud.google.com/optimization/docs
.. _Setup Authentication.: https://googleapis.dev/python/google-api-core/latest/auth.html

Installation
~~~~~~~~~~~~

Install this library in a `virtualenv`_ using pip. `virtualenv`_ is a tool to
create isolated Python environments. The basic problem it addresses is one of
dependencies and versions, and indirectly permissions.

With `virtualenv`_, it's possible to install this library without needing system
install permissions, and without clashing with the installed system
dependencies.

.. _`virtualenv`: https://virtualenv.pypa.io/en/latest/


Code samples and snippets
~~~~~~~~~~~~~~~~~~~~~~~~~

Code samples and snippets live in the `samples/` folder.


Supported Python Versions
^^^^^^^^^^^^^^^^^^^^^^^^^
Our client libraries are compatible with all current `active`_ and `maintenance`_ versions of
Python.

Python >= 3.7

.. _active: https://devguide.python.org/devcycle/#in-development-main-branch
.. _maintenance: https://devguide.python.org/devcycle/#maintenance-branches

Unsupported Python Versions
^^^^^^^^^^^^^^^^^^^^^^^^^^^
Python <= 3.6

If you are using an `end-of-life`_
version of Python, we recommend that you update as soon as possible to an actively supported version.

.. _end-of-life: https://devguide.python.org/devcycle/#end-of-life-branches

Mac/Linux
^^^^^^^^^

.. code-block:: console

    pip install virtualenv
    virtualenv <your-env>
    source <your-env>/bin/activate
    <your-env>/bin/pip install google-cloud-optimization


Windows
^^^^^^^

.. code-block:: console

    pip install virtualenv
    virtualenv <your-env>
    <your-env>\Scripts\activate
    <your-env>\Scripts\pip.exe install google-cloud-optimization

Next Steps
~~~~~~~~~~

-  Read the `Client Library Documentation`_ for Cloud Optimization API
   to see other available methods on the client.
-  Read the `Cloud Optimization API Product documentation`_ to learn
   more about the product and see How-to Guides.
-  View this `README`_ to see the full list of Cloud
   APIs that we cover.

.. _Cloud Optimization API Product documentation:  https://cloud.google.com/optimization/docs
.. _README: https://github.com/googleapis/google-cloud-python/blob/main/README.rst


CHANGELOG.md
# Changelog

## [1.1.0](https://github.com/googleapis/python-optimization/compare/v1.0.1...v1.1.0) (2022-08-02)


### Features

* add audience parameter ([2b17d6a](https://github.com/googleapis/python-optimization/commit/2b17d6a68205a46a31d0bf5e1bbe62af8f98ac66))


### Bug Fixes

* **deps:** require google-api-core>=1.32.0,>=2.8.0 ([#48](https://github.com/googleapis/python-optimization/issues/48)) ([2b17d6a](https://github.com/googleapis/python-optimization/commit/2b17d6a68205a46a31d0bf5e1bbe62af8f98ac66))
* require python 3.7+ ([#50](https://github.com/googleapis/python-optimization/issues/50)) ([a49598b](https://github.com/googleapis/python-optimization/commit/a49598b350b3d4688e8742d6371027a028915073))

## [1.0.1](https://github.com/googleapis/python-optimization/compare/v1.0.0...v1.0.1) (2022-06-07)


### Bug Fixes

* **deps:** require protobuf>=3.19.0,<4.0.0 ([#42](https://github.com/googleapis/python-optimization/issues/42)) ([c141479](https://github.com/googleapis/python-optimization/commit/c141479451116a9667c73e4dc391e2933048da0c))


### Documentation

* fix changelog header to consistent size ([#40](https://github.com/googleapis/python-optimization/issues/40)) ([d1feb93](https://github.com/googleapis/python-optimization/commit/d1feb935cbe4ed15588adf165988cc74fcdb0a73))

## [1.0.0](https://github.com/googleapis/python-optimization/compare/v0.1.1...v1.0.0) (2022-04-26)


### Features

* bump release level to production/stable ([#31](https://github.com/googleapis/python-optimization/issues/31)) ([dd9377f](https://github.com/googleapis/python-optimization/commit/dd9377f24b3e1ffcaee7c23c72872696bbaa6a20))

## [0.1.1](https://github.com/googleapis/python-optimization/compare/v0.1.0...v0.1.1) (2022-04-25)


### Documentation

* add code snippets for async api ([#18](https://github.com/googleapis/python-optimization/issues/18)) ([60b35a0](https://github.com/googleapis/python-optimization/commit/60b35a01b12d2c0034aebe1edcc25487bd0fe567))
* add get_operation code snippets ([#12](https://github.com/googleapis/python-optimization/issues/12)) ([a95c19f](https://github.com/googleapis/python-optimization/commit/a95c19fef17c86f587febcf054a7f1fa49851cdf))
* add long timeout code snippet ([#20](https://github.com/googleapis/python-optimization/issues/20)) ([9f57450](https://github.com/googleapis/python-optimization/commit/9f574507010ef637e5a6912a1cb725b782c03cf4))
* add sync api samples with json request ([#13](https://github.com/googleapis/python-optimization/issues/13)) ([c3d6087](https://github.com/googleapis/python-optimization/commit/c3d60874977628698c7f3d0b137c120971e7c42c))
* update operation id ([#23](https://github.com/googleapis/python-optimization/issues/23)) ([d9c6c42](https://github.com/googleapis/python-optimization/commit/d9c6c422d6146f65f11fa98370f9b7f7edd166ad))

## 0.1.0 (2022-03-24)


### Features

* generate v1 ([122d1da](https://github.com/googleapis/python-optimization/commit/122d1da807b2637b9cebb43a4df4f01cbe9feef2))


owl-bot-staging/v1/noxfile.py
# -*- coding: utf-8 -*-
# Copyright 2022 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
import os
import pathlib
import shutil
import subprocess
import sys


import nox  # type: ignore

ALL_PYTHON = [
    "3.6",
    "3.7",
    "3.8",
    "3.9",
    "3.10",
]

CURRENT_DIRECTORY = pathlib.Path(__file__).parent.absolute()

LOWER_BOUND_CONSTRAINTS_FILE = CURRENT_DIRECTORY / "constraints.txt"
PACKAGE_NAME = subprocess.check_output([sys.executable, "setup.py", "--name"], encoding="utf-8")

BLACK_VERSION = "black==19.10b0"
BLACK_PATHS = ["docs", "google", "tests", "samples", "noxfile.py", "setup.py"]
DEFAULT_PYTHON_VERSION = "3.9"

nox.sessions = [
    "unit",
    "cover",
    "mypy",
    "check_lower_bounds"
    # exclude update_lower_bounds from default
    "docs",
    "blacken",
    "lint",
    "lint_setup_py",
]

@nox.session(python=ALL_PYTHON)
def unit(session):
    """Run the unit test suite."""

    session.install('coverage', 'pytest', 'pytest-cov', 'pytest-asyncio', 'asyncmock; python_version < "3.8"')
    session.install('-e', '.')

    session.run(
        'py.test',
        '--quiet',
        '--cov=google/cloud/optimization_v1/',
        '--cov=tests/',
        '--cov-config=.coveragerc',
        '--cov-report=term',
        '--cov-report=html',
        os.path.join('tests', 'unit', ''.join(session.posargs))
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def cover(session):
    """Run the final coverage report.
    This outputs the coverage report aggregating coverage from the unit
    test runs (not system test runs), and then erases coverage data.
    """
    session.install("coverage", "pytest-cov")
    session.run("coverage", "report", "--show-missing", "--fail-under=100")

    session.run("coverage", "erase")


@nox.session(python=ALL_PYTHON)
def mypy(session):
    """Run the type checker."""
    session.install('mypy', 'types-pkg_resources')
    session.install('.')
    session.run(
        'mypy',
        '--explicit-package-bases',
        'google',
    )


@nox.session
def update_lower_bounds(session):
    """Update lower bounds in constraints.txt to match setup.py"""
    session.install('google-cloud-testutils')
    session.install('.')

    session.run(
        'lower-bound-checker',
        'update',
        '--package-name',
        PACKAGE_NAME,
        '--constraints-file',
        str(LOWER_BOUND_CONSTRAINTS_FILE),
    )


@nox.session
def check_lower_bounds(session):
    """Check lower bounds in setup.py are reflected in constraints file"""
    session.install('google-cloud-testutils')
    session.install('.')

    session.run(
        'lower-bound-checker',
        'check',
        '--package-name',
        PACKAGE_NAME,
        '--constraints-file',
        str(LOWER_BOUND_CONSTRAINTS_FILE),
    )

@nox.session(python=DEFAULT_PYTHON_VERSION)
def docs(session):
    """Build the docs for this library."""

    session.install("-e", ".")
    session.install("sphinx==4.0.1", "alabaster", "recommonmark")

    shutil.rmtree(os.path.join("docs", "_build"), ignore_errors=True)
    session.run(
        "sphinx-build",
        "-W",  # warnings as errors
        "-T",  # show full traceback on exception
        "-N",  # no colors
        "-b",
        "html",
        "-d",
        os.path.join("docs", "_build", "doctrees", ""),
        os.path.join("docs", ""),
        os.path.join("docs", "_build", "html", ""),
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def lint(session):
    """Run linters.

    Returns a failure if the linters find linting errors or sufficiently
    serious code quality issues.
    """
    session.install("flake8", BLACK_VERSION)
    session.run(
        "black",
        "--check",
        *BLACK_PATHS,
    )
    session.run("flake8", "google", "tests", "samples")


@nox.session(python=DEFAULT_PYTHON_VERSION)
def blacken(session):
    """Run black. Format code to uniform standard."""
    session.install(BLACK_VERSION)
    session.run(
        "black",
        *BLACK_PATHS,
    )


@nox.session(python=DEFAULT_PYTHON_VERSION)
def lint_setup_py(session):
    """Verify that setup.py is valid (including RST check)."""
    session.install("docutils", "pygments")
    session.run("python", "setup.py", "check", "--restructuredtext", "--strict")
